### PR TITLE
MG-1961 - Add Domains OpenAPI Spec

### LINKED ISSUE #253: Feature: Missing Domain API spec
### Is your feature request related to a problem? Please describe.

Missing Domain API spec

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Add Domain API spec 

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by arvindh123**
# What type of PR is this?
This is a documentation update because it updates the following documentation: 
- Add Domains openAPI spec 

## What does this do?
- Add Domains openAPI spec 

## Which issue(s) does this PR fix/relate to?

- Resolves absmach/magistrala#1961 

## Have you included tests for your changes?

No,  Not Applicable

## Did you document any new/modified feature?

No, Not Applicable

### CODE REVIEW

File: api/openapi/auth.yml [hunk 1]
```diff
@@ -32,6 +32,305 @@ tags:
...
+        $ref: "#/components/requestBodies/DomainCreateReq"
+      responses:
+        "201":
+          $ref: "#/components/responses/DomainCreateRes"
+        "400":
+          description: Failed due to malformed JSON.
+        "401":
+          description: Missing or invalid access token provided.
+        "409":
+          description: Failed due to using an existing identity.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:52]: the unique variable being used at the moment is `alias`?

File: api/openapi/auth.yml [hunk 2]
```diff
@@ -32,6 +32,305 @@ tags:
...
+        "422":
+          description: Database can't process request.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+
+  /domains/{domainID}:
+    get:
+      summary: Retrieves domain information
+      description: |
+        Retrieves a specific domain that is identifier by the domain ID.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:90]:
```suggestion
        Retrieves a specific domain that is identified by the domain ID.
```

File: api/openapi/auth.yml [hunk 3]
```diff
@@ -32,6 +32,305 @@ tags:
...
+          description: A non-existent entity request.
+        "422":
+          description: Database can't process request.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+
+    patch:
+      summary: Updates name and alias of the domain.
+      description: |
+        Updates name and alias of the domain.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:112]: `metadata` and `tags` are also updated in this request

File: api/openapi/auth.yml [hunk 4]
```diff
@@ -32,6 +32,305 @@ tags:
...
+        "403":
+          description: Unauthorized access to domain id.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+
+  /domains/{domainID}/enable:
+    post:
+      summary: Enables a domain
+      description: |
+        Enables a specific domain that is identifier by the domain ID.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:161]:
```suggestion
        Enables a specific domain that is identified by the domain ID.
```

File: api/openapi/auth.yml [hunk 5]
```diff
@@ -32,6 +32,305 @@ tags:
...
+        "422":
+          description: Database can't process request.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+
+  /domains/{domainID}/disable:
+    post:
+      summary: Disable a domain
+      description: |
+        Disable a specific domain that is identifier by the domain ID.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:188]:
```suggestion
        Disable a specific domain that is identified by the domain ID.
```

File: api/openapi/auth.yml [hunk 6]
```diff
@@ -32,6 +32,305 @@ tags:
...
+        "422":
+          description: Database can't process request.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+
+  /domains/{domainID}/freeze:
+    post:
+      summary: Freeze a domain
+      description: |
+        Freeze a specific domain that is identifier by the domain ID.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:215]:
```suggestion
        Freeze a specific domain that is identified by the domain ID.
```

Comment (ianmuchyri) [reply api/openapi/auth.yml:215]: do the same for the rest

File: api/openapi/auth.yml [hunk 7]
```diff
@@ -32,6 +32,305 @@ tags:
...
+          description: List of users assigned to domain.
+        "400":
+          description: Failed due to malformed domain's ID.
+        "401":
+          description: Missing or invalid access token provided.
+        "403":
+          description: Unauthorized access the domain ID.
+        "404":
+          description: A non-existent entity request.
+        "409":
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:328]: does a get request result in an error `409`?

File: api/openapi/auth.yml [hunk 8]
```diff
@@ -150,6 +449,156 @@ paths:
...
+        status:
+          type: string
+          description: Domain Status
+          format: string
+          example: enabled
+        created_by:
+          type: string
+          format: uuid
+          example: "0d837f56-3f8a-4e2a-9359-6347d0fc9f06 "
+          description: User ID of user created domain.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:513]:
```suggestion
          description: User ID of the user who created the domain.
```

File: api/openapi/auth.yml [hunk 9]
```diff
@@ -150,6 +449,156 @@ paths:
...
+        created_at:
+          type: string
+          format: date-time
+          example: "2019-11-26 13:31:52"
+          description: Time when the domain was created.
+        updated_by:
+          type: string
+          format: uuid
+          example: "80f66b77-ed74-4e74-9f88-6cce9a0a3049"
+          description: User ID of user last updated domain.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:523]:
```suggestion
          description: User ID of the user who last updated the domain.
```

File: api/openapi/auth.yml [hunk 10]
```diff
@@ -150,6 +449,156 @@ paths:
...
+        updated_at:
+          type: string
+          format: date-time
+          example: "2019-11-26 13:31:52"
+          description: Time when the domain was last update.
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:528]:
```suggestion
          description: Time when the domain was last updated.
```

File: api/openapi/auth.yml [hunk 11]
```diff
@@ -259,6 +742,36 @@ components:
...
   requestBodies:
+    DomainCreateReq:
+      description: JSON-formatted document describing the new domain to be registered
+      required: true
+      content:
+        application/json:
+          schema:
+            $ref: "#/components/schemas/DomainReqObj"
+    DomainUpdateReq:
+      description: JSON-formated document describing the metadata and name of domain to be update
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:753]:
```suggestion
      description: JSON-formated document describing the name, alias, tags, and metadata of the domain to be updated
```

File: api/openapi/auth.yml [hunk 12]
```diff
@@ -259,6 +742,36 @@ components:
...
+    AssignUserReq:
+      description: JSON-formated document describing the policy related to assigning users to a domain
+      required: true
+      content:
+        application/json:
+          schema:
+            $ref: "#/components/schemas/UserDomainRelationReq"
+
+    UnassignUsersReq:
+      description: JSON-formated document describing the policy related to assigning users to a domain
```
Comment (ianmuchyri) [inline api/openapi/auth.yml:768]:
```suggestion
      description: JSON-formated document describing the policy related to unassigning users to a domain
```

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

File: api/openapi/auth.yml (hunk 1 above)
Comment (arvindh123) [reply api/openapi/auth.yml:52]: Yes

File: api/openapi/auth.yml (hunk 3 above)
Comment (arvindh123) [reply api/openapi/auth.yml:112]: Yes , I will update the description

File: api/openapi/auth.yml (hunk 1 above)
Comment (ianmuchyri) [reply api/openapi/auth.yml:52]: this then should be:

```suggestion
          description: Failed due to using an existing alias.
```

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

**Review by ianmuchyri — APPROVED** [review]

File: api/openapi/auth.yml [hunk 13]
```diff
@@ -32,6 +32,303 @@ tags:
       url: http://docs.mainflux.io/
 
 paths:
+  /domains:
```
Comment (rodneyosodo) [inline api/openapi/auth.yml:35]: Also document 
- list_domains_by_user_id
- view_domain_permissions

File: api/openapi/auth.yml [hunk 14]
```diff
@@ -32,6 +32,303 @@ tags:
...
+          description: Unauthorized access the domain ID.
+        "404":
+          description: A non-existent entity request.
+        "409":
+          description: Conflict of data.
+        "422":
+          description: Database can't process request.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+  /domains/{domainID}/users:
```
Comment (rodneyosodo) [inline api/openapi/auth.yml:299]: This endpoint is not implemented

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: api/openapi/auth.yml (hunk 13 above)
Comment (arvindh123) [reply api/openapi/auth.yml:35]: /users/{userID}/domains - This endpoint belongs to users , but endpoint is present in Auth service

File: api/openapi/auth.yml (hunk 14 above)
Comment (arvindh123) [reply api/openapi/auth.yml:299]: This endpoint was implemented in Users service

Comment (arvindh123) [reply api/openapi/auth.yml:299]: It was implemented in Users service ,

Comment (rodneyosodo) [reply api/openapi/auth.yml:299]: Shouldn't it be documented in the users service?

File: api/openapi/auth.yml (hunk 13 above)
Comment (rodneyosodo) [reply api/openapi/auth.yml:35]: But isn't it under auth service?

Comment (arvindh123) [reply api/openapi/auth.yml:35]: In the OpenAPI specification, I suggest structuring API endpoints not based on service-specific endpoints but rather on entities. 
For example, within the `users.yml` file, all endpoints would start with `/users`. Operations related to specific users, such as retrieving user tags (`users/{userID}/tag`), would be included here. However, I propose including additional endpoints, like `/users/{userID}/things`, which pertain to the 'things' associated with a user, even if the 'things' functionality resides in a different service.

By including such endpoints in the `users.yml` API specification, users interacting with this document will find it easier to understand as they would expect `users.Yaml` to provide comprehensive user-related functionalities, even if certain functionalities are fulfilled by other services.

Comment (rodneyosodo) [reply api/openapi/auth.yml:35]: Shouldn't it be documentation of a specific micro-service(server)?

Comment (dborovcanin) [reply api/openapi/auth.yml:35]: I do not mind having endpoints from different services, but we generally keep services and entities logically correlated anyway. For the given example, maybe we should change the API for `Things`. I.e. this is filtering of things, so something like `/things?user=id` is more suitable anyways.

Comment (arvindh123) [reply api/openapi/auth.yml:35]: I agree with your idea https://github.com/absmach/magistrala-old/pull/261#discussion_r1447229367
I will check the possibilities in code and create a new issue

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**